File: 2021_11.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include "include/grid.h"


#ifdef DEBUG
#define DBGPRINT(...) fprintf(stderr, __VA_ARGS__)
#else
#define DBGPRINT(...)
#endif

#define AOC_DAY 10
#define AOC_YEAR 2021

int add_energy(struct grid * g, int x, int y)
{
    int value;

    DBGPRINT("adding energy (%d,%d)\n", x, y);

    /* returns 1 if octopus flashes right now; else 0 */
    value = grid_get_value(g, x, y) + 1;
    grid_set_value(g, x, y, value);
    if(10 == value)
    {
        /* this octopus just flashed! */
        return 1;
    }

    return 0;
}

void energize_adjacent(struct grid * g, int x, int y)
{
    int tmpx, tmpy;
    /* energizing all connected octopi (including diagonals) */
    DBGPRINT("ENERGIZE ADJACENT (%d,%d)\n", x, y);
    tmpx = x - 1;
    if(tmpx >= 0)
    {
        /* left column */
        for(tmpy = y - 1; tmpy <= y + 1; tmpy++)
        {
            if(tmpy >= 0 && tmpy < g->height)
            {
                if(add_energy(g, tmpx, tmpy))
                    /* this octopus just flashed! */
                    energize_adjacent(g, tmpx, tmpy);
            }
        }
    }
    /* top */
    if(y - 1 >= 0)
    {
        if(add_energy(g, x, y - 1))
            /* this octopus just flashed! */
            energize_adjacent(g, x, y - 1);
    }
    if(y + 1 < g->height)
    {
        if(add_energy(g, x, y + 1))
            /* this octopus just flashed! */
            energize_adjacent(g, x, y + 1);
    }
    /* bottom */
    tmpx = x + 1;
    if(tmpx < g->width)
    {
        /* right column */
        for(tmpy = y - 1; tmpy <= y + 1; tmpy++)
        {
            if(tmpy >= 0 && tmpy < g->height)
            {
                if(add_energy(g, tmpx, tmpy))
                    /* this octopus just flashed! */
                    energize_adjacent(g, tmpx, tmpy);
            }
        }
    }
    DBGPRINT("  DONE ENERGIZE ADJACENT (%d,%d)\n", x, y);
}

unsigned long energize_octopi(struct grid * g)
{
    unsigned long nflashes = 0;
    int x, y, value;

    /* add 1 energy to all octopi, and if flash, energize adjacent */
    for(x=0; x < g->width; x++)
    {
        for(y=0; y < g->height; y++)
        {
            value = grid_get_value(g, x, y) + 1;
            grid_set_value(g, x, y, value);
            if(10 == value)
            {
                energize_adjacent(g, x, y);
            }
        }
    }

    /* set every octopus that was a 10 back to 0 */
    for(x=0; x < g->width; x++)
    {
        for(y=0; y < g->height; y++)
        {
            if(grid_get_value(g, x, y) >= 10)
            {
                nflashes++;
                grid_set_value(g, x, y, 0);
            }
        }
    }

    return nflashes;
}

int main(int argc, char ** argv)
{
    struct grid g;
    unsigned long nflashes = 0, nflashes_this_step;
    int first_total_flash = -1;

    read_grid(&g, NULL);

    for(int step = 0; step < 100 || -1 == first_total_flash; step++)
    {
        nflashes_this_step = energize_octopi(&g);
        if(step < 100)
            nflashes += nflashes_this_step;
        printf("%d (need %d)\n", nflashes_this_step, g.width * g.height);
        if((-1 == first_total_flash) \
                && (nflashes_this_step == g.width * g.height))
        {
            first_total_flash = step;
        }
    }

    printf("After 100 steps: %lu flashes\n", nflashes);
    printf("First total flash: %d\n", first_total_flash + 1);

    free_grid(&g);
    return 0;
}

File: 2021_11.rb
require 'logger'
require_relative 'lib/grid'

AOCDAY = 11
AOCYEAR = 2021

def flash(grid)
  flashstack = []
  (0...grid.width).each do |x|
    (0...grid.height).each do |y|
      # increment energy level for this octopus
      energylevel = grid.get(x, y)
      grid.set(x, y, energylevel + 1)

      # push every octopus that needs to be flashed on the flashstack
      if energylevel >= 9
        flashstack << [x,y]
      end
    end
  end

  # pop each flashing octopus. Add any octopi this flash causes to flash
  # to the flashstack.
  flasher = flashstack.pop
  while not flasher.nil?
    x,y = flasher
    # increment the energy level of every octopus adjacent to this octopus
    [[x-1,y+1], [x,y+1], [x+1,y+1], [x-1,y], [x+1,y], [x-1,y-1], [x,y-1], [x+1,y-1]].each do |neighborx, neighbory|
      # if this is an octopus, increment its energy level
      begin
        energylevel = grid.get(neighborx,neighbory)
        grid.set(neighborx, neighbory, energylevel + 1)

        # if this octopus flashes, add it to the flashstack
        if 9 == energylevel
          flashstack << [neighborx, neighbory]
        end
      rescue GridBoundaryError => err
      end
    end
    flasher = flashstack.pop
  end

  # process all flashes (value 10 in the grid)
  nflashes = 0
  (0...grid.width).each do |x|
    (0...grid.height).each do |y|
      if grid.get(x,y) > 9
        nflashes += 1
        grid.set(x, y, 0)
      end
    end
  end

  # return number of flashes
  nflashes
end

def run(inputfile, logger, **kwargs)
  # read grid
  grid = Grid.new(fh: File.open(inputfile), single_character: true)

  flashcount = 0;
  nsteps = 0;
  
  flashcount_this_step = 0;
  # loop ends after all octopi have flashed together
  while flashcount_this_step != grid.width * grid.height
    # count flashes this step
    flashcount_this_step = flash(grid);

    # record total number of flashes in first 100 steps
    nsteps += 1;
    if nsteps <= 100
      flashcount += flashcount_this_step
    end
    if nsteps < 10
      logger.debug("Step #{nsteps}:\n#{grid.to_s}")
    end
  end

  puts("#{flashcount} flashes in 100 steps!");
  puts("All octopi flashed together at step #{nsteps}");
end


if __FILE__ == $0
  require 'optparse'

  options = {}
  OptionParser.new do |opts|
    opts.banner = "Usage: $0 [options] <inputfile> [<more inputfiles>]\n" \
      "  Advent of Code #{AOCYEAR} day #{AOCDAY}"

    opts.on("-v", "--verbose", "Be verbose") do |v|
      options[:verbose] = v
    end

    opts.on("-2", "--part2", "Run part 2") do |part2|
      options[:part2] = true
    end

    opts.on("-h", "--help", "Print this help message") do
      puts opts
      exit
    end
  end.parse!

  logger = Logger.new(STDOUT);
  logger.level = (options.include? :verbose) ? Logger::DEBUG : Logger::WARN

  ARGV.each do |inputfile|
    run(inputfile, logger, **options)
  end
end
